contract/models/contract_line.py

```diff
@@ -580,23 +601,32 @@ def _insert_markers(self, first_date_invoiced, last_date_invoiced):
     @api.multi
     def _update_recurring_next_date(self):
         for rec in self:
-            old_date = rec.recurring_next_date
-            new_date = old_date + self.get_relative_delta(
-                rec.recurring_rule_type, rec.recurring_interval
-            )
-            if rec.recurring_rule_type == 'monthlylastday':
-                last_date_invoiced = old_date
-            elif rec.recurring_invoicing_type == 'post-paid':
-                last_date_invoiced = old_date - relativedelta(days=1)
-            elif rec.recurring_invoicing_type == 'pre-paid':
-                last_date_invoiced = new_date - relativedelta(days=1)
-
-            if rec.date_end and last_date_invoiced >= rec.date_end:
-                rec.last_date_invoiced = rec.date_end
-                rec.recurring_next_date = False
-            else:
-                rec.last_date_invoiced = last_date_invoiced
-                rec.recurring_next_date = new_date
+            if rec.next_period_end_date:
+                last_date_invoiced = rec.next_period_end_date
+                next_period_start_date = last_date_invoiced + relativedelta(
+                    days=1
+                )
+                next_period_end_date = self._get_next_period_end_date(
+                    next_period_start_date,
+                    rec.recurring_rule_type,
+                    rec.recurring_interval,
+                )
+                recurring_next_date = self._get_recurring_next_date(
+                    next_period_start_date,
+                    rec.recurring_invoicing_type,
+                    rec.recurring_rule_type,
+                    rec.recurring_interval,
+                )
+                if rec.date_end and next_period_end_date > rec.date_end:
+                    next_period_end_date = rec.date_end
+                if rec.date_end and last_date_invoiced >= rec.date_end:
+                    rec.last_date_invoiced = rec.date_end
+                    rec.recurring_next_date = False
+                    rec.next_period_end_date = False
+                else:
+                    rec.last_date_invoiced = last_date_invoiced
+                    rec.recurring_next_date = recurring_next_date
+                    rec.next_period_end_date = next_period_end_date
```

Comment:
I think you can continue simplifying by doing `_update_next_period_date` instead of `_update_recurring_next_date` and make `recurring_next_date` a computed field that depends on `next_period_date` and `recurring_invoicing_type`.